internal/policy/branch_restrictor.go

```diff
@@ -0,0 +1,104 @@
+package policy
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"regexp/syntax"
+
+	"github.com/lunarway/release-manager/internal/git"
+	"github.com/lunarway/release-manager/internal/log"
+	"github.com/pkg/errors"
+)
+
+type BranchRestrictor struct {
+	ID            string `json:"id,omitempty"`
+	BranchMatcher string `json:"branchMatcher,omitempty"`
+	Environment   string `json:"environment,omitempty"`
+}
+
+// ApplyBranchRestrictor applies a branch-restrictor policy for service svc to
+// environment env with regular expression branchMatcher.
+func (s *Service) ApplyBranchRestrictor(ctx context.Context, actor Actor, svc, branchMatcher, env string) (string, error) {
+	span, ctx := s.Tracer.FromCtx(ctx, "policy.ApplyBranchRestrictor")
+	defer span.Finish()
+
+	// validate branch matcher regular expression before storring
+	_, err := syntax.Parse(branchMatcher, syntax.Perl)
+	if err != nil {
+		return "", errors.WithMessage(err, "branch matcher not valid")
+	}
+
+	commitMsg := git.PolicyUpdateApplyCommitMessage(env, svc, "branch-restrictor")
```

Comment:
Should we validate if an auto-release policy exists that will not be allowed by this?

Reply:
That would be a really good use experience. Is it noted somewhere?